glusterd2/commands/snapshot/snapshot-restore.go

```diff
@@ -0,0 +1,375 @@
+	}
+
+	return nil
+}
+
+func registerSnapRestoreStepFuncs() {
+	var sfs = []struct {
+		name string
+		sf   transaction.StepFunc
+	}{
+		{"snap-restore.Commit", snapRestore},
+		{"snap-restore.UndoCommit", undoSnapRestore},
+		{"snap-restore.UndoStore", undoSnapStore},
+		{"snap-restore.Store", storeSnapRestore},
+	}
+	for _, sf := range sfs {
+		transaction.RegisterStepFunc(sf.sf, sf.name)
+	}
+}
+
+func snapshotRestoreHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+	snapname := mux.Vars(r)["snapname"]
+
+	snapinfo, err := snapshot.GetSnapshot(snapname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+
+	txn, err := transaction.NewTxnWithLocks(ctx, snapname, snapinfo.ParentVolume)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	defer txn.Done()
+
+	//Fetching snapinfo again, but this time inside a lock
+	snapinfo, err = snapshot.GetSnapshot(snapname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	snapvolinfo := &snapinfo.SnapVolinfo
+
+	vol, err := volume.GetVolume(snapinfo.ParentVolume)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	if vol.State == volume.VolStarted {
+		errMsg := fmt.Sprintf("Volume %s must be in stopped state before restoring.", vol.Name)
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errMsg)
+		return
+	}
+
+	txn.Steps = []*transaction.Step{
+		{
+			DoFunc:   "snap-restore.Commit",
+			UndoFunc: "snap-restore.UndoCommit",
+			Nodes:    snapvolinfo.Nodes(),
+		},
+		{
+			DoFunc:   "snap-restore.Store",
+			UndoFunc: "snap-restore.UndoStore",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+		},
+	}
+	err = txn.Ctx.Set("snapname", snapname)
+	if err != nil {
+		logger.WithError(err).Error("failed to set request in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+	err = txn.Ctx.Set("snapinfo", snapinfo)
+	if err != nil {
+		logger.WithError(err).Error("failed to set request in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Ctx.Set("volinfo", vol)
+	if err != nil {
+		logger.WithError(err).Error("failed to set request in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Do()
+	if err != nil {
+		logger.WithError(err).Error("snapshot restore transaction failed")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	msg := fmt.Sprintf("Snapshot %s restored to volume %s", snapvolinfo.Name, vol.Name)
+	txn.Ctx.Logger().WithField("Snap Name", snapname).Info(msg)
```

Comment:
Avoid space in key name. Change this to "snap-name"

Reply:
You've used "snapshot" elsewhere as the key for snapname, so use the same here.